### DISCUSSION TITLE
Add Custom Token Support for testing

### DISCUSSION THREAD

**Discussion by kwdikeman** [Synthetixio/synpress#359] [Q&A]
Does/has anyone extended the metamask commands to support the adding of custom tokens (add-token)?

I'm starting this task but am getting hung at the input of the add-token page.  What am I missing?

What I've done so far: 

1.  Added a new page (add-custom-token)
`// Add Custom Token Page

const importTokensPage='.add-token';
const input=`${importTokensPage} #custom-address`;
const inputTokenName=`${importTokensPage} #custom-symbol`;
const inputTokenDecimals=`${importTokensPage} #custom-decimals`;
const cancelButton=`${importTokensPage} .btn-default`;
const nextTokenPage=`${importTokensPage} .btn-primary`;

module.exports.tokenPageElements = {
    input,
    inputTokenName,
    inputTokenDecimals,
    cancelButton,
    nextTokenPage,
};

// Confirm Add Token Page
const confirmImportTokenPage='.confirm-add-token';
const confirmTokenAdd=`${confirmImportTokenPage} .btn-secondary`;

module.exports.confirmTokenPageElements = {
    confirmTokenAdd,
}`

2.  Added code to command.js
  `// Add Custom Token
Cypress.Commands.add('addCustomToken', (tokenAddress) => {
  return cy.task('addCustomToken', {tokenAddress}
  );
});`

3.  Added code to plugins.js
`addCustomToken: async tokenAddress => {
      const added = await metamask.importToken(tokenAddress);
      return added;
    },`

4. Added code to metamask.js
`// Add Custom Token
  importToken: async tokenAddress => {
    
    await switchToMetamaskIfNotActive();
    await puppeteer.waitFor(mainPageElements.importToken.button);
    if (tokenAddress) {
      await puppeteer.waitAndType(tokenPageElements.input, tokenAddress);
      await puppeteer.waitAndClick(tokenPageElements.nextTokenPage);
      await puppeteer.waitFor(confirmTokenPageElements.confirmTokenAdd)
      await puppeteer.waitAndClick(confirmTokenPageElements.confirmTokenAdd);
    }  
    return true;
  },`

**Comment by drptbl** [comment]
Hey @kwdikeman, I will implement adding a custom token this week. Cheers!

**Reply by drptbl** [reply]
will keep you updated 👍

**Comment by kwdikeman** [comment]
I got it to work.  I'm not sure why but by creating a new page and using
that descriptor it could not see the form elements.  I pulled the elements
out and treated it as if the main page was the target.

KD
